packages/expo-modules-core/ios/JSI/EXJavaScriptValue.mm

```diff
@@ -0,0 +1,161 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+#import <ExpoModulesCore/EXJSIConversions.h>
+#import <ExpoModulesCore/EXJavaScriptValue.h>
+#import <ExpoModulesCore/EXJavaScriptRuntime.h>
+
+@implementation EXJavaScriptValue {
+  __weak EXJavaScriptRuntime *_runtime;
+  std::shared_ptr<jsi::Value> _value;
+}
+
+- (nonnull instancetype)initWithRuntime:(nonnull EXJavaScriptRuntime *)runtime
+                                  value:(std::shared_ptr<jsi::Value>)value
+{
+  if (self = [super init]) {
+    _runtime = runtime;
+    _value = value;
+  }
+  return self;
+}
+
+#pragma mark - Type checking
+
+- (BOOL)isUndefined
+{
+  return _value->isUndefined();
+}
+
+- (BOOL)isNull
+{
+  return _value->isNull();
+}
+
+- (BOOL)isBool
+{
+  return _value->isBool();
+}
+
+- (BOOL)isNumber
+{
+  return _value->isNumber();
+}
+
+- (BOOL)isString
+{
+  return _value->isString();
+}
+
+- (BOOL)isSymbol
+{
+  return _value->isSymbol();
+}
+
+- (BOOL)isObject
+{
+  return _value->isObject();
+}
+
+- (BOOL)isFunction
+{
+  if (_value->isObject()) {
+    jsi::Runtime *runtime = [_runtime get];
+    return _value->getObject(*runtime).isFunction(*runtime);
+  }
+  return false;
+}
+
++ (nonnull NSString *)kindOf:(nonnull EXJavaScriptValue *)value
+{
+  if ([value isUndefined]) {
+    return @"undefined";
+  }
+  if ([value isNull]) {
+    return @"null";
+  }
+  if ([value isBool]) {
+    return @"boolean";
+  }
+  if ([value isNumber]) {
+    return @"number";
+  }
+  if ([value isString]) {
+    return @"string";
+  }
+  if ([value isFunction]) {
+    return @"function";
+  }
+  assert([value isObject] && "Expecting object.");
```

Comment:
```suggestion
  assert([value isObject], "Expecting object.");
```
I think that "&&" won't work here.

Reply:
It works 😄 Actually I copied this from JSI. `assert` itself is single-argument macro and has no option for a custom message, so people usually do it that way.